File: trees_and_graphs/binary_search_tree_checker/binary_search_tree_checker.js
class BinaryTreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }

  insertLeft(value) {
    this.left = new BinaryTreeNode(value);
    return this.left;
  }

  insertRight(value) {
    this.right = new BinaryTreeNode(value);
    return this.right;
  }
}

function isBinarySearchTree(treeRoot) {

  // Determine if the tree is a valid binary search tree
  // doing bfs is okay but at each node, I need to keep track of a lowerBound and upperBound
  // except at root
  // i.e. root.left (its lowerBound is nothing, upperBound is root.value)
  // every subsequent child's lowerbound for nodes to the right is the parent, 
  // every child's upperbound is the parent if it is to the left
  // so yes I can do BFS but might as well do dfs because the depth stored should be O(lgn)
  const nodes = [[treeRoot, undefined, undefined]];
  
  while (nodes.length > 0) {
    const currNodeInfo = nodes.pop();
    const currNode = currNodeInfo[0];
    const currLowerBound = currNodeInfo[1];
    const currUpperBound = currNodeInfo[2];
    
    if (currNode.left) {
      // child of this node
      // add it to array along with its lower and upper bound
      const leftNode = currNode.left;
      // this value should be less than currNode value
      if (leftNode.value > currNode.value || (currLowerBound && leftNode.value < currLowerBound)) { return false; }
      // upper bound is currNode
      // lower bound is currNode's lower bound
      const leftNodeInfo = [leftNode, currLowerBound, currNode.value];
      nodes.push(leftNodeInfo);
    }

    if (currNode.right) {
      const rightNode = currNode.right;
      // i must be bigger than this lower bound (parent node) and less than the upper bound which I get from parent
      if (rightNode.value < currNode.value || (currUpperBound && rightNode.value > currUpperBound)) { return false; }

      // lower bound is currNode value, upper bound is curr node's upper bound
      const rightNodeInfo = [rightNode, currNode.value, currUpperBound];
      nodes.push(rightNodeInfo);
    }
  }

  // checked the entire tree's values and nothing was wrong
  return true;
}

treeRoot = new BinaryTreeNode(50);
leftNode = treeRoot.insertLeft(30);
leftNode.insertLeft(20);
leftNode.insertRight(60);
rightNode = treeRoot.insertRight(80);
rightNode.insertLeft(70);
rightNode.insertRight(90);
isBinarySearchTree(treeRoot);

File: trees_and_graphs/binary_search_tree_checker/binary_search_tree_checker.py
class BinaryTreeNode(object):

  def __init__(self, value):
    self.value = value
    self.left = None
    self.right = None

  def insert_left(self, value):
    self.left = BinaryTreeNode(value)
    return self.left

  def insert_right(self, value):
    self.right = BinaryTreeNode(value)
    return self.right

def is_binary_search_tree(root):

  # Determine if the tree is a valid binary search tree
  # Using DFS is better if the tree is balanced since a depth will be O(lgn)
  # First I need to set up a stack containing my root
  # While the stack is not empty, the last in node will be popped
  # and I will examine if it is a valid node, aka if it has any left or right nodes, that each one obeys 
  # its boundaries

  # The tuple will contain the lower_bound and upper_bound at the first and second indices for the node
  # Root node has no lower and upper bound so I placed None there
  tree_stack = [(root, None, None)]

  while len(tree_stack):
    curr_node, lower_bound, upper_bound = tree_stack.pop()

    if curr_node.left:
      # check that left node is a valid node
      # left node value should be less than upper bound which is curr_node value
      # AND greater than the lower bound given by curr_node
      left_node = curr_node.left
      if left_node.value > curr_node.value or (lower_bound and left_node.value < lower_bound):
        return False
      # passed condition, so add child node to the stack
      tree_stack.append((left_node, lower_bound, curr_node.value))
    
    if curr_node.right:
      #same as above, check that right node is valid
      # Right node value should be > than lower bound aka curr_node value 
      # AND lesser than the upper bound given by curr_node
      right_node = curr_node.right 
      if right_node.value < curr_node.value or (upper_bound and right_node.value > upper_bound):
        return False
      tree_stack.append((right_node, curr_node.value, upper_bound))
  
  return True


